### PR TITLE
Batch Resource Management via MQTT (Very Alpha!!) 

### PR DESCRIPTION
**Pull request by RickCarlino**
# Disclaimer

**Huge, alpha-grade changes ahead. API may change (or disappear) without notice. Use caution or abstain from writing applications that rely on this feature!**

# Why?

FarmBot uses two APIs: one for controlling devices (MQTT) and another for controlling resources (HTTP).

 * Some users have mentioned interest in controlling all aspects of a device via a single API.
 * Some users have requested the ability to perform batch API operations in a single API request.

# What's New?

This PR introduces an experimental API for managing resources over MQTT. Currently, only `"delete"` actions are supported, but additional actions will be added in a future (next few weeks) PR as the feature is tested in real-world use cases.

# How Does it Work?

See `batch_updates.md`.

# What's Next?

 * Lot's and lots of QA.
 * Support batch `"update"` actions.
 * Support batch `"create"` actions.

CC: @creimers , @phrohdoh

### CODE REVIEW

File: app/lib/resources/preprocessor.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+module Resources
+  # Takes a bunch of unsafe, string-y data that came in over AMQP and parses it
+  # into fully formed
+  class PreProcessor < Mutations::Command
+    def self.from_amqp(delivery_info, body)
+      # ["bot", "device_3", "resources_v0", "destroy", "Sequence", "2", "123-456
```
Comment (Phrohdoh) [inline app/lib/resources/preprocessor.rb]: This comment seems unfinished?

File: app/mutations/create_destroyer.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# I heard you like mutations. So we made a mutation class that creates mutations
+# so you can mutate while you mutate.
+# This class will create a "base case" `::Delete` mutation. Very useful when you
```
[the remark this answers is not in the document]
Comment (RickCarlino) [reply app/mutations/create_destroyer.rb]: Good catch. We used to have `::Delete` mutations and I occasionally find myself using the old format.

File: batch_updates.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+```
+bot/device_3/resources_v0/destroy/Sequence/2/123-456
+```
+
+NOTES:
+
+ * `<Transaction UUID>` can be any user defined string. Ensure that the string is unique. We recommend using UUIDs.
+ * `<resource_id>` This is the `.id` property of the resource you are deleting.
+ * `<action>` Only `delete` is supported as of July 2018.
```
[the remark this answers is not in the document]
Comment (RickCarlino) [reply batch_updates.md]: Yep, that's a legit typo. Will fix in the next commit- I've been debugging some issues with the service worker on staging.

File: app/lib/resources/preprocessor.rb (hunk 1 above)
Comment (RickCarlino) [reply app/lib/resources/preprocessor.rb]: That was a "note to self" about what `delivery_info.routing_key.split(".")` looks like at runtime.

Here's a more in-depth explanation:

```ruby
      # -- Snip! --
      # Parse the AMQP rotuing key into an Array of strings.
      # A properly formatted routing_key will look like this after processing:
      # 
      # ["bot", "device_3", "resources_v0", "destroy", "Sequence", "2", "xyz"]
      segments = delivery_info.routing_key.split(".")
      _, device_name, _, action, resource, resource_id, uuid = segments
     # -- Snip! --
```

**Pull request merged**